lib/IRGen/IRGenSIL.cpp

```diff
@@ -1987,26 +1987,31 @@ void IRGenSILFunction::visitGraphOperationInst(GraphOperationInst *i) {
          "could not find TensorArrayProtocol protocol");
   assert(outputTensorGroupProto && "could not find TensorGroup protocol");
 
-  if (!llvm::TFDynamicCompilation) {
-    // If we are not in dynamic compilation mode, then deabstraction may not
-    // have transformed the graph_ops into a form that we can lower, so do not
-    // try to lower the graph_ops.
-    // TODO(marcrasi): Once we make deabstraction unnecessary for IRGen, we
-    // should be able to lower everything no matter what the mode is, so remove
-    // this check.
-    const std::string errMessage = "!!! Compiler bug -- graph_op " +
-                              opInfo.getOperationName().str() +
-                              " cannot be lowered to LLVM IR !!!\n";
-    abortOnGraphOp(*this, errMessage.c_str());
-
-    // Finally, we set up our explosion results full of undef values.
+  // Sets the results of this op to undef.
+  auto lowerResultsToUndef = [&]() {
     for (auto result : i->getResults()) {
       ExplosionSchema schema = getTypeInfo(result->getType()).getSchema();
       Explosion e;
       for (auto &elt : schema)
         e.add(llvm::UndefValue::get(elt.getScalarType()));
       setLoweredExplosion(result, e);
     }
+  };
+
+  // Makes this op cause a runtime error.
+  auto lowerOpToError = [&](const char *errorMessage) {
+    abortOnGraphOp(*this, errorMessage);
+    lowerResultsToUndef();
+  };
+
+  // These ops configure the `deviceInfo`. They do not do anything
+  // at runtime, so do not lower them.
+  // This must be above the GraphFunctionDeviceInfo::getForFunction() call,
+  // because GraphFunctionDeviceInfo::getForFunction() crashes when called in
+  // the `enableTPU` function, because the enableInfeed operand isn't an integer
+  // literal instruction.
+  if (GraphFunctionDeviceInfo::isConfigOp(opInfo)) {
+    lowerResultsToUndef();
```

Comment:
it seems lowerResultsToUndef is used in two cases:
1. when we have generated a runtime error, so that the lowered undef is guaranteed to not execute. the callsite above does this.
2. when we want to ignore the lowering of a certain inst, so that running that inst is a no-op. the callsite here does this.

for #2, there is a risk that if anyone later consumes the result, it could crash (or cause UB?). would it be better to instead assert that in these cases, there are 0 results?

Reply:
yes, asserting sounds better than lowering to undef. they actually have one Void result, so I asserted that they have 1 result and then set the result to the Void value.